### PR TITLE
[CPF-3147] EasyDAO-TransactionDAO

### PR DESCRIPTION
**Pull request by jlhughes**

### CODE REVIEW

File: src/foam/nanos/ruler/predicate/PropertyEQValue.js [hunk 1]
```diff
@@ -40,11 +40,12 @@ foam.CLASS({
     {
       name: 'f',
       javaCode: `
-        if ( getIsNew() ) {
+        FObject old = (FObject) OLD_OBJ.f(obj);
+        if ( getIsNew() ||
+             old == null ) {
```
Comment (lchanmann) [inline src/foam/nanos/ruler/predicate/PropertyEQValue.js]: when isNew=false, we want to compare the oldObj.prop to the value and if the oldObj is null should we just return false? @nanoMichal 

(similar thing for ropertyNEQValue.js and PropertyEQProperty.js)

File: src/foam/nanos/ruler/Rule.js [hunk 2]
```diff
@@ -275,9 +275,8 @@
           try {
             return getPredicate().f(obj);
           } catch ( Throwable th ) { }
-
-          ((Logger) x.get("logger")).error(
-            "Failed to evaluate predicate of rule: " + getId(), t);
+          // ((Logger) x.get("logger")).debug(this.getClass().getSimpleName(), "id", getId(), "\\nrule", this, "\\nobj", obj, "\\nold", oldObj, "\\n", t);
```
Comment (lchanmann) [inline src/foam/nanos/ruler/Rule.js:278]: do we want to leave the commented code here?

File: src/foam/nanos/ruler/RuleEngine.java [hunk 3]
```diff
@@ -57,21 +57,27 @@ public void setDelegate(DAO delegate) {
...
+        if ( ! isRuleApplicable(rule, obj, oldObj)) continue;
+        PM pm = (PM) x_.get("PM");
+        pm.setClassType(RulerDAO.getOwnClassInfo());
+        pm.setName(rule.getDaoKey() + ": " + rule.getId());
+        pm.init_();
+        applyRule(rule, obj, oldObj, agency);
+        pm.log(x_);
+        agency.submit(x_, x -> saveHistory(rule, obj), "Save history. Rule id:" + rule.getId());
+      } catch ( Exception e ) {
+        // logger.debug(this.getClass().getSimpleName(), "id", rule.getId(), "\\nrule", rule, "\\nobj", obj, "\\nold", oldObj, "\\n", e);
```
Comment (lchanmann) [inline src/foam/nanos/ruler/RuleEngine.java:73]: same as above

**Review by lchanmann — CHANGES_REQUESTED** [review]

File: src/foam/nanos/ruler/Rule.js (hunk 2 above)
Comment (jlhughes) [reply src/foam/nanos/ruler/Rule.js:278]: yes. debugging rules is tricky, and keep enabling these to see what is going on.

File: src/foam/nanos/ruler/predicate/PropertyEQValue.js (hunk 1 above)
Comment (jlhughes) [reply src/foam/nanos/ruler/predicate/PropertyEQValue.js]: without the check for old == null we have NPE because the old tests are blindly executed.

Comment (lchanmann) [reply src/foam/nanos/ruler/predicate/PropertyEQValue.js]: we can add the null check before the last return statement.
```
return old != null && EQ(old.getClassInfo().getAxiomByName ...
```

**Review by lchanmann — APPROVED** [review]

**Pull request merged**